Reduce unsafeness in encryptedmedia module

annevk opened the pull request:
#### 71f279c32f7c40a816392098cd8c61024974b408
<pre>
Reduce unsafeness in encryptedmedia module
<a href="https://bugs.webkit.org/show_bug.cgi?id=295462">https://bugs.webkit.org/show_bug.cgi?id=295462</a>

Reviewed by Chris Dumez and Rupin Mittal.

Helps with <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

Canonical link: <a href="https://commits.webkit.org/297058@main">https://commits.webkit.org/297058@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9cde7170c2557244818ee219f08211180f9ceb1d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110415 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30074 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20507 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116440 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60669 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112378 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30753 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38661 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83964 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113363 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24540 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99422 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64405 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23902 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17561 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60236 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93923 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17617 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119231 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37455 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27789 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92931 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37828 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95690 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92754 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37752 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15487 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33401 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17809 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37350 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42821 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37012 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40352 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38720 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebCore/Modules/encryptedmedia/legacy/WebKitMediaKeySession.cpp:
```diff
@@ -198,8 +201,7 @@ void WebKitMediaKeySession::addKeyTimerFired()
 
-            ASSERT(m_keys);
-            m_keys->keyAdded();
+            Ref { *m_keys }->keyAdded();
```
I think we should keep the assert so we crash safely in the event it's nullptr.

annevk replied:
Crashing on nullptr is a prerequisite for Safer C++ and we already do that.

cdumez replied:
> Crashing on nullptr is a prerequisite for Safer C++ and we already do that.

Uh, I do not think this is generally true. Dereferencing nullptr is undefined behavior and we have not solved this problem yet (because it regresses perf). Some smart pointers will RELEASE_ASSERT when referencing though but you should check, not assume. Here m_keys is a raw pointer so no check when doing `*` on it. Then `Ref`'s `operator->()` only has a debug assert, not a release assert.

Given the debug ASSERT in Ref, dropping the assert outside may be fine but I am not 100% sure. The compiler is free to optimize things and assume non-null when you pass a reference and not a pointer I believe. Ideally we want a RELEASE_ASSERT for safety though.

RupinMittal commented on Source/WebCore/Modules/encryptedmedia/legacy/LegacyCDMSessionClearKey.cpp:
```diff
@@ -64,7 +64,7 @@ RefPtr<Uint8Array> CDMSessionClearKey::generateKeyRequest(const String& mimeType
 
     bool sawError = false;
-    String keyID = PAL::UTF8Encoding().decode(m_initData->span(), true, sawError);
+    String keyID = PAL::UTF8Encoding().decode(initData->span(), true, sawError);
```
Maybe I'm missing something--what's the purpose of this change?

annevk replied:
This is a trick we also employ elsewhere. m_initData isn't protected, but initData has to be protected by the caller and the analyzer knows that. So sometimes we use the argument instead of the member variable to avoid it being labeled as unsafe.

RupinMittal reviewed (approved):
Left two small comments. Otherwise, LGTM!

cdumez reviewed (approved):
LGTM but I think we want a RELEASE_ASSERT in WebKitMediaKeySession for safety. Actually, even better, `m_keys` would be a `CheckedPtr` since a raw pointer data member is not safe and since `CheckedPtr` operator will do the release assertion for you:
```
    ALWAYS_INLINE T& operator*() const { RELEASE_ASSERT(m_ptr); return *get(); }
    ALWAYS_INLINE T* operator->() const { RELEASE_ASSERT(m_ptr); return get(); }
```

annevk committed:
Reduce unsafeness in encryptedmedia module
https://bugs.webkit.org/show_bug.cgi?id=295462

Reviewed by Chris Dumez and Rupin Mittal.

Helps with https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

Canonical link: https://commits.webkit.org/297058@main
